### ISSUE TITLE
error: invalid initialization of non-const reference of type ‘nvinfer1::ILogger&’ from an rvalue of type ‘nvinfer1::ILogger*’

### ISSUE THREAD

**Issue by huihui308** [marcoslucianops/DeepStream-Yolo#368]
There is a error when I compile nvdsinfer_custom_impl_Yolo. The platform I running is nvidia xaiver, jpack is 4.6.
```
$ CUDA_VER=10.2 make -C nvdsinfer_custom_impl_Yolo
make: Entering directory '/home/lcfc/david/code/DeepStream-Yolo/nvdsinfer_custom_impl_Yolo'
g++ -c  -o yolo.o -Wall -std=c++11 -shared -fPIC -Wno-error=deprecated-declarations -I/opt/nvidia/deepstream/deepstream/sources/includes -I/usr/local/cuda-10.2/include yolo.cpp
yolo.cpp: In member function ‘nvinfer1::ICudaEngine* Yolo::createEngine(nvinfer1::IBuilder*, nvinfer1::IBuilderConfig*)’:
yolo.cpp:80:60: error: invalid initialization of non-const reference of type ‘nvinfer1::ILogger&’ from an rvalue of type ‘nvinfer1::ILogger*’
     parser = nvonnxparser::createParser(*network, getLogger());
                                                   ~~~~~~~~~^~
In file included from yolo.cpp:26:0:
/usr/include/aarch64-linux-gnu/NvOnnxParser.h:242:17: note: in passing argument 2 of ‘nvonnxparser::IParser* nvonnxparser::{anonymous}::createParser(nvinfer1::INetworkDefinition&, nvinfer1::ILogger&)’
 inline IParser* createParser(nvinfer1::INetworkDefinition& network, nvinfer1::ILogger& logger)
                 ^~~~~~~~~~~~
yolo.cpp:207:83: warning: ‘nvinfer1::ICudaEngine* nvinfer1::IBuilder::buildEngineWithConfig(nvinfer1::INetworkDefinition&, nvinfer1::IBuilderConfig&)’ is deprecated [-Wdeprecated-declarations]
   nvinfer1::ICudaEngine* engine = builder->buildEngineWithConfig(*network, *config);
                                                                                   ^
In file included from /usr/include/aarch64-linux-gnu/NvOnnxParser.h:26:0,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:7990:43: note: declared here
     TRT_DEPRECATED nvinfer1::ICudaEngine* buildEngineWithConfig(
                                           ^~~~~~~~~~~~~~~~~~~~~
Makefile:81: recipe for target 'yolo.o' failed
make: *** [yolo.o] Error 1
```

**Comment by huihui308** [comment]
> Deepstream is 6.0. The platform I running is nvidia xaiver, jpack is 4.6.

**Comment by marcoslucianops** [comment]
Did you change the `yolo.cpp` file?

This line

```
parser = nvonnxparser::createParser(*network, getLogger());
```

**Comment by huihui308** [comment]
> Did you change the `yolo.cpp` file?
> 
> This line
> 
> ```
> parser = nvonnxparser::createParser(*network, getLogger());
> ```
> 
> It should be
> 
> ```
> parser = nvonnxparser::createParser(*network, *builder->getLogger());
> ```

No, I did not. The log as followings.
```
$ CUDA_VER=10.2 make -C nvdsinfer_custom_impl_Yolo
make: Entering directory '/home/lcfc/david/code/DeepStream-Yolo/nvdsinfer_custom_impl_Yolo'
g++ -c  -o yolo.o -Wall -std=c++11 -shared -fPIC -Wno-error=deprecated-declarations -I/opt/nvidia/deepstream/deepstream/sources/includes -I/usr/local/cuda-10.2/include yolo.cpp
yolo.cpp: In member function ‘nvinfer1::ICudaEngine* Yolo::createEngine(nvinfer1::IBuilder*, nvinfer1::IBuilderConfig*)’:
yolo.cpp:80:61: error: ‘class nvinfer1::IBuilder’ has no member named ‘getLogger’
     parser = nvonnxparser::createParser(*network, *builder->getLogger());
                                                             ^~~~~~~~~
yolo.cpp:207:83: warning: ‘nvinfer1::ICudaEngine* nvinfer1::IBuilder::buildEngineWithConfig(nvinfer1::INetworkDefinition&, nvinfer1::IBuilderConfig&)’ is deprecated [-Wdeprecated-declarations]
   nvinfer1::ICudaEngine* engine = builder->buildEngineWithConfig(*network, *config);
                                                                                   ^
In file included from /usr/include/aarch64-linux-gnu/NvOnnxParser.h:26:0,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:7990:43: note: declared here
     TRT_DEPRECATED nvinfer1::ICudaEngine* buildEngineWithConfig(
                                           ^~~~~~~~~~~~~~~~~~~~~
Makefile:81: recipe for target 'yolo.o' failed
make: *** [yolo.o] Error 1
make: Leaving directory '/home/lcfc/david/code/DeepStream-Yolo/nvdsinfer_custom_impl_Yolo'
```

**Comment by huihui308** [comment]
> Did you change the `yolo.cpp` file?
> 
> This line
> 
> ```
> parser = nvonnxparser::createParser(*network, getLogger());
> ```
> 
> It should be
> 
> ```
> parser = nvonnxparser::createParser(*network, *builder->getLogger());
> ```

I have solved this issue by modifing:
```
parser = nvonnxparser::createParser(*network, *gTrtLogger);
```
And add gTrtLogger definition out of this function.
```
#include <memory>
std::unique_ptr<nvinfer1::ILogger> gTrtLogger;
```
And then I can compile the lib success.

But there is a segment fault when running 
```
deepstream-app -c deepstream_app_config.txt 
ERROR: Deserialize engine failed because file path: /home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine open error
0:00:01.403649278 28065   0x55b6121ec0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1889> [UID = 1]: deserialize engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed
0:00:01.403802406 28065   0x55b6121ec0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:1996> [UID = 1]: deserialize backend context from engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed, try rebuild
0:00:01.403860297 28065   0x55b6121ec0 INFO                 nvinfer gstnvinfer.cpp:647:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1914> [UID = 1]: Trying to create engine from model files
Segmentation fault (core dumped)
```
I use gdb to debug this issu, the information is as follows.
```
$ gdb deepstream-app
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from deepstream-app...(no debugging symbols found)...done.
(gdb) r -c deepstream_app_config.txt
Starting program: /usr/bin/deepstream-app -c deepstream_app_config.txt
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
[New Thread 0x7f86fb3b50 (LWP 28129)]
[New Thread 0x7f867b2b50 (LWP 28130)]
[New Thread 0x7f84a13b50 (LWP 28131)]
[New Thread 0x7f81aa5b50 (LWP 28132)]
[New Thread 0x7f812a4b50 (LWP 28133)]
ERROR: Deserialize engine failed because file path: /home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine open error
0:00:02.112514303 28125   0x5555b9f6c0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1889> [UID = 1]: deserialize engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed
0:00:02.112652870 28125   0x5555b9f6c0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:1996> [UID = 1]: deserialize backend context from engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed, try rebuild
0:00:02.112737674 28125   0x5555b9f6c0 INFO                 nvinfer gstnvinfer.cpp:647:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1914> [UID = 1]: Trying to create engine from model files

Thread 1 "deepstream-app" received signal SIGSEGV, Segmentation fault.
0x0000007f82244e48 in ?? () from /usr/lib/aarch64-linux-gnu/libnvonnxparser.so.8
(gdb) where
#0  0x0000007f82244e48 in  () at /usr/lib/aarch64-linux-gnu/libnvonnxparser.so.8
#1  0x0000007f8092c474 in Yolo::createEngine(nvinfer1::IBuilder*, nvinfer1::IBuilderConfig*) ()
    at /home/lcfc/david/code/DeepStream-Yolo/nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so
#2  0x0000007f8093c71c in NvDsInferYoloCudaEngineGet () at /home/lcfc/david/code/DeepStream-Yolo/nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so
#3  0x0000007f83f56e74 in nvdsinfer::TrtModelBuilder::getCudaEngineFromCustomLib(bool (*)(nvinfer1::IBuilder*, _NvDsInferContextInitParams*, nvinfer1::DataType, nvinfer1::ICudaEngine*&), bool (*)(nvinfer1::IBuilder*, nvinfer1::IBuilderConfig*, _NvDsInferContextInitParams const*, nvinfer1::DataType, nvinfer1::ICudaEngine*&), _NvDsInferContextInitParams const&, NvDsInferNetworkMode&) () at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#4  0x0000007f83f57254 in nvdsinfer::TrtModelBuilder::buildModel(_NvDsInferContextInitParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) () at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#5  0x0000007f83f12938 in nvdsinfer::NvDsInferContextImpl::buildModel(_NvDsInferContextInitParams&) () at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#6  0x0000007f83f13404 in nvdsinfer::NvDsInferContextImpl::generateBackendContext(_NvDsInferContextInitParams&) ()
    at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#7  0x0000007f83f0eb34 in nvdsinfer::NvDsInferContextImpl::initialize(_NvDsInferContextInitParams&, void*, void (*)(INvDsInferContext*, unsigned int, NvDsInferLogLevel, char const*, void*)) () at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#8  0x0000007f83f13cb8 in createNvDsInferContext(INvDsInferContext**, _NvDsInferContextInitParams&, void*, void (*)(INvDsInferContext*, unsigned int, NvDsInferLogLevel, char const*, void*)) () at /opt/nvidia/deepstream/deepstream-6.0/lib/libnvds_infer.so
#9  0x0000007f84036144 in gst_nvinfer_start(_GstBaseTransform*) () at /usr/lib/aarch64-linux-gnu/gstreamer-1.0/deepstream/libnvdsgst_infer.so
#10 0x0000007fab9aa224 in  () at /usr/lib/aarch64-linux-gnu/libgstbase-1.0.so.0
#11 0x0000005555b6a620 in  ()
(gdb) 
```

How can I solved this problem?
I run deepstream-6.0 on xavier jetson, the jpack is 4.6.

**Comment by huihui308** [comment]
> > Did you change the `yolo.cpp` file?
> > This line
> > ```
> > parser = nvonnxparser::createParser(*network, getLogger());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It should be
> > ```
> > parser = nvonnxparser::createParser(*network, *builder->getLogger());
> > ```
> 
> I have solved this issue by modifing:
> 
> ```
> Logger logger;
> parser = nvonnxparser::createParser(*network, logger.getTRTLogger());
> ```
> 
> And add Logger definition out of this function.
> 
> ```
> class Logger : public nvinfer1::ILogger
> {
> public:
> 	Logger(Severity severity = Severity::kWARNING)
> 	{
> 		severity = severity;
> 	}
> 
> 	~Logger() = default;
> 
> 	nvinfer1::ILogger& getTRTLogger()
> 	{
> 		return *this;
> 	}
> 
>     void log(nvinfer1::ILogger::Severity severity, const char* msg) noexcept override
>     {
>         // suppress info-level messages
>         if (severity == Severity::kINFO) return;
> 
>         switch (severity)
>         {
>         case Severity::kINTERNAL_ERROR: std::cerr << "INTERNAL_ERROR: " << msg << std::endl; break;
> 		case Severity::kERROR: std::cerr << "ERROR: " << msg << std::endl; break;
>         case Severity::kWARNING: std::cerr << "WARNING: " << msg << std::endl; break;
>         case Severity::kINFO: std::cerr << "INFO: " << msg << std::endl; break;
>         case Severity::kVERBOSE: break;
>       //  default: std::cerr <<"UNKNOW:"<< msg << std::endl;break;
>         }
>     }
> };
> ```
> 
> And then I can compile the lib success.
> 
> But there is another issue when I running.
> 
> ```
> $ deepstream-app -c deepstream_app_config.txt                            
> ERROR: Deserialize engine failed because file path: /home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine open error
> 0:00:01.238592781 31493   0x557cc062c0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1889> [UID = 1]: deserialize engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed
> 0:00:01.238723411 31493   0x557cc062c0 WARN                 nvinfer gstnvinfer.cpp:644:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:1996> [UID = 1]: deserialize backend context from engine from file :/home/lcfc/david/code/DeepStream-Yolo/model_b1_gpu0_fp32.engine failed, try rebuild
> 0:00:01.238756373 31493   0x557cc062c0 INFO                 nvinfer gstnvinfer.cpp:647:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1914> [UID = 1]: Trying to create engine from model files
> WARNING: onnx2trt_utils.cpp:364: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
> ERROR: ModelImporter.cpp:720: While parsing node number 239 [Range -> "/0/model.22/Range_output_0"]:
> ERROR: ModelImporter.cpp:721: --- Begin node ---
> ERROR: ModelImporter.cpp:722: input: "/0/model.22/Constant_8_output_0"
> input: "/0/model.22/Cast_output_0"
> input: "/0/model.22/Constant_9_output_0"
> output: "/0/model.22/Range_output_0"
> name: "/0/model.22/Range"
> op_type: "Range"
> 
> ERROR: ModelImporter.cpp:723: --- End node ---
> ERROR: ModelImporter.cpp:726: ERROR: builtin_op_importers.cpp:3172 In function importRange:
> [8] Assertion failed: inputs.at(0).isInt32() && "For range operator with dynamic inputs, this version of TensorRT only supports INT32!"
> 
> Could not parse the ONNX model
> 
> Failed to build CUDA engine
> ERROR: Failed to create network using custom network creation function
> ERROR: Failed to get cuda engine from custom library API
> 0:00:01.513890898 31493   0x557cc062c0 ERROR                nvinfer gstnvinfer.cpp:641:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Error in NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1934> [UID = 1]: build engine file failed
> 0:00:01.514005976 31493   0x557cc062c0 ERROR                nvinfer gstnvinfer.cpp:641:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Error in NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:2020> [UID = 1]: build backend context failed
> 0:00:01.514071931 31493   0x557cc062c0 ERROR                nvinfer gstnvinfer.cpp:641:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Error in NvDsInferContextImpl::initialize() <nvdsinfer_context_impl.cpp:1257> [UID = 1]: generate backend failed, check config file settings
> 0:00:01.514197730 31493   0x557cc062c0 WARN                 nvinfer gstnvinfer.cpp:850:gst_nvinfer_start:<primary_gie> error: Failed to create NvDsInferContext instance
> 0:00:01.514263813 31493   0x557cc062c0 WARN                 nvinfer gstnvinfer.cpp:850:gst_nvinfer_start:<primary_gie> error: Config file path: /home/lcfc/david/code/DeepStream-Yolo/config_infer_primary_yoloV8.txt, NvDsInfer Error: NVDSINFER_CONFIG_FAILED
> ** ERROR: <main:679>: Failed to set pipeline to PAUSED
> Quitting
> ERROR from primary_gie: Failed to create NvDsInferContext instance
> Debug info: gstnvinfer.cpp(850): gst_nvinfer_start (): /GstPipeline:pipeline/GstBin:primary_gie_bin/GstNvInfer:primary_gie:
> Config file path: /home/lcfc/david/code/DeepStream-Yolo/config_infer_primary_yoloV8.txt, NvDsInfer Error: NVDSINFER_CONFIG_FAILED
> App run failed
> ```
> 
> How can I solved this problem? I run deepstream-6.0 on xavier jetson, the jpack is 4.6.

I have solved this problem. I generate onnx file using this command.
```
python3 export_yoloV8.py -w best.pt --dynamic -s 640
```
Then the above issue occured. but when I removed '--dynamic' and the issue disappeared.
```
python3 export_yoloV8.py -w best.pt -s 640
```

**Comment by huihui308** [comment]
> I will fix the `logger` issue in the next update. About the second error, the version of the DeepStream (specifically the TensorRT version) you are using doesn't support `--dynamic` flag (due to `INT64` weigths). Set the `--batch` in the ONNX exporter equal to the `batch-size` you will use on the DeepStream.

Thank you. I appreciate it.
